test/async-hooks/test-embedder.api.async-resource.js

```diff
@@ -9,9 +9,21 @@ const { AsyncResource } = async_hooks;
 const initHooks = require('./init-hooks');
 const { checkInvocations } = require('./hook-checks');
 
+// Verify that if there is no registered hook, then those invalid parameters
+// won't be checked.
+assert.doesNotThrow(() => new AsyncResource());
```

Comment:
I screwed up the logic here. It shouldn't return early before the arguments are checked. Otherwise it'll cause an async hook stack corruption. Here's the minimal test case:
```js
new (require('async_hooks').AsyncResource)().emitBefore();
```
Here's the diff for the fix. Feel free to add it to this PR along with the above test case:
```diff
diff --git a/lib/async_hooks.js b/lib/async_hooks.js
index d5d5407..aec73ed 100644
--- a/lib/async_hooks.js
+++ b/lib/async_hooks.js
@@ -205,15 +205,15 @@ class AsyncResource {
     }
     this[trigger_id_symbol] = triggerId;
 
-    // Return immediately if there's nothing to do.
-    if (async_hook_fields[kInit] === 0)
-      return;
-
     if (typeof type !== 'string' || type.length <= 0)
       throw new TypeError('type must be a string with length > 0');
     if (!Number.isSafeInteger(triggerId) || triggerId < 0)
       throw new RangeError('triggerId must be an unsigned integer');
 
+    // Return immediately if there's nothing to do.
+    if (async_hook_fields[kInit] === 0)
+      return;
+
     processing_hook = true;
     for (var i = 0; i < active_hooks_array.length; i++) {
       if (typeof active_hooks_array[i][init_symbol] === 'function') {
```